xds/internal/xdsclient/authority_test.go

```diff
@@ -0,0 +1,344 @@
+	a, mgmtServer, nodeID := setupAuthorityWithMgmtServer(ctx, t, mgmtServerOpts, defaultTestTimeout)
+	defer a.close()
+	defer mgmtServer.Stop()
+
+	resourceName := "xdsclient-test-lds-resource"
+	watcher := newTestResourceWatcher()
+	cancelResource := a.watchResource(listenerResourceType, resourceName, watcher)
+	defer cancelResource()
+
+	// Waiting for mgmt server to recv request before verifying state.
+	select {
+	case <-ctx.Done():
+		t.Fatal("Test timed out before mgmt server got the request.")
+	case <-onStreamRequestDoneCh:
+	}
+	if err := compareWStateAndWTimer(a, listenerResourceType, resourceName, watchStateRequested); err != nil {
+		t.Fatal(err)
+	}
+
+	resources := e2e.UpdateOptions{
+		NodeID:         nodeID,
+		Listeners:      []*v3listenerpb.Listener{e2e.DefaultClientListener(resourceName, "new-rds-resource")},
+		SkipValidation: true,
+	}
+	if err := mgmtServer.Update(ctx, resources); err != nil {
+		t.Fatalf("Failed to update management server with resources: %v, err: %v", resources, err)
+	}
+
+	// Waiting for authority to receive the update before verifying state.
+	select {
+	case <-ctx.Done():
+		t.Fatal("Test timed out before watcher received the update.")
+	case <-watcher.updateCh:
+	}
+
+	if err := compareWStateAndWTimer(a, listenerResourceType, resourceName, watchStateReceived); err != nil {
+		t.Fatal(err)
+	}
+}
+
+// This tests the resource's watch state transition when the ADS stream is closed
+// by the management server. The test calls the `watchResource` api to register
+// a watch for a resource, stops the management server, and verifies the resource's
+// watch state transitions from `watchStateRequested` to `watchStateStarted` so
+// that the watch can be restarted later.
+func (s) TestResourceStateTransitionsFromRequestedToStartedOnError(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	onStreamRequestDoneCh := make(chan struct{})
+	mgmtServerOpts := mgmtServerOptWithOnRequestDoneCh(onStreamRequestDoneCh)
+	a, mgmtServer, _ := setupAuthorityWithMgmtServer(ctx, t, mgmtServerOpts, defaultTestTimeout)
+	defer a.close()
+
+	resourceName := "xdsclient-test-lds-resource"
+	watcher := newTestResourceWatcher()
+	cancelResource := a.watchResource(listenerResourceType, resourceName, watcher)
+	defer cancelResource()
+
+	// Waiting for mgmt server to recv request.
+	select {
+	case <-ctx.Done():
+		t.Fatal("Test timed out before mgmt server got the request.")
+	case <-onStreamRequestDoneCh:
+	}
+	if err := compareWStateAndWTimer(a, listenerResourceType, resourceName, watchStateRequested); err != nil {
+		t.Fatal(err)
+	}
+
+	mgmtServer.Stop()
+
+	// Waiting for watcher to receive the update.
+	select {
+	case <-ctx.Done():
+		t.Fatal("Test timed out before verifying error propagation.")
+	case err := <-watcher.errorCh:
+		if err == nil {
+			t.Fatal("got err == nil. Want: err from stream connection reset by peer")
```

Comment:
Prefer a slightly more readable error string. Maybe `Connection error not propagated to watchers"`

Also, is it possible (or required) to verify that the actual error received by the watcher is the expected one?